enhancements/alerting/alerting-as-a-feature.md

```diff
@@ -0,0 +1,302 @@
+---
+title: Alerting as a feature
+authors:
+  - "@dofinn"
+reviewers:
+  - "@smarterclayton"
+  - "@jeremyeder"
+  - "@michaelgugino"
+  - "@s-urbaniak"
+  - "@lilic"
+  - "@RiRa12621"
+  - "@wking"
+  - "@cblecker"
+  - "@jharrington22"
+  - "@mwoodson"
+approvers:
+  - TBD
+creation-date: 2021-03-08
+last-updated: 2021-03-08
+status: implementable
+---
+
+# Alerting as a Feature
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [x] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Definitions
+
+Caused Based Alert: Describes the direct cause of an issue. Example: etcdMemoryUtilization @ 100%
+
+Symptom Based Alert = Describes a symptom who's source is a cause based alert. Example: etcd connection latency > 200ms. This may be caused by etcdMemoryUtilization (or not?)
```

Comment:
How high up the stack you go before a cause-based alert becomes a symptom-based alert depends on your particular SLAs and SLOs.  [The Prom docs][1] link Rob Ewaschuk's excellent [philosophy doc][2], talking about [alerting from the spout][3].  If you're managing etcd, etcd request latency is a reasonable spout.  If you're managing the cluster, where folks can only get at etcd through the Kube API servers, etcd latency becomes a cause-based alert explaining one possible component of your spout Kube API latency alert.

Besides case-based and symptom/spout alerting, I think it's also useful to have diagnostics that look at firing alerts and other in-cluster status and say "ah, the cluster of issues you're seeing look like `$ROOT_CAUSE`".  The [insights folks][4] have [some tools in this space][5], but currently they are internal and don't run in-cluster.

[1]: https://prometheus.io/docs/practices/alerting/
[2]: https://docs.google.com/document/d/199PqyG3UsyXlwieHaqbGiWVa8eMWi8zzAn0YfcApr8Q/edit
[3]: https://docs.google.com/document/d/199PqyG3UsyXlwieHaqbGiWVa8eMWi8zzAn0YfcApr8Q/edit#heading=h.rvrk9gcasjzh
[4]: https://github.com/openshift/insights-operator
[5]: https://www.openshift.com/blog/openshift-insights-for-openshift-cluster-manager

Reply:
> How high up the stack you go before a cause-based alert becomes a symptom-based

Start from as a high up the stack as you can go then. We should be alerting from the perspective of outside in, not inside out.

Reply:
Side note, does this separation of alerts matter in the overall proposal? I think it's fair to say that alerting should be a feature of Openshift yes, but the content or alerts could be of a choice of cluster admins really, plus it depends what is being run and with what SLA.

The reason is that cause/symptom alert statements are controversial as noticed. As you mentioned the write up about this from Google SRE here: https://docs.google.com/document/d/199PqyG3UsyXlwieHaqbGiWVa8eMWi8zzAn0YfcApr8Q/edit#heading=h.6ammb5h32uqq is excellent material.
The rule is simple:  `don't write cause-based paging rules for symptoms you can catch otherwise.`

We need to also identify the goal of alerting here. It is not for drilling down (!). It's for notifying you that users are or will be  impacted(!)

> That said, if your debugging dashboards let you move quickly enough from symptom to cause to amelioration, you don't need to spend time on cause-based rules anyway.

Reply:
+1 to the above thoughts. Thats why i proposed alerting profiles to enable admins/SRSs to either:
a) configure alerting from scratch
b) configure alerting with caused based alerts (as a starting point)
c) configure alerting with symptom based alerts (as a starting point)